Add documentation about the compositor (#9409)
godotengine/godot-docs: tutorials/rendering/compositor.rst

.. _doc_compositor:

The Compositor
==============

The compositor is a new feature in Godot 4 that allows control over
the rendering pipeline used when rendering the contents of a viewport.

It can be configured on a :ref:`WorldEnvironment <class_WorldEnvironment>`
node where it applies to all viewports, or it can be configured on
a :ref:`Camera3D <class_Camera3D>` and apply only to
the viewport using that camera.

The :ref:`Compositor <class_Compositor>` resource is used to configure

> the compositor, to get started, simply create a new compositor on

Reviewer: ```suggestion
the compositor. To get started, simply create a new compositor on
```

the appropriate node:

.. image:: img/new_compositor.webp

.. note::
    The compositor is currently a feature that is only supported for

>     the Mobile and Forward+ renderer.

Reviewer: ```suggestion
    the Mobile and Forward+ renderers.
```

Compositor effects
------------------

Compositor effects allow you to insert additional logic into the rendering
pipeline at various stages. This is an advanced feature that requires
a high level of understanding of the rendering pipeline to use to
its best advantage.

As the core logic of the compositor effect is called from the rendering
pipeline it is important to note that this logic will thus run within
the thread on which rendering takes place.
Care needs to be taken to ensure we don't run into threading issues.

To illustrate how to use compositor effects we'll create a simple
post processing effect that allows you to write your own shader code
and apply this full screen through a compute shader.
You can find the finished demo project `here <https://github.com/godotengine/godot-demo-projects/tree/master/compute/post_shader>`_.

We start by creating a new script called ``post_process_shader.gd``.

> We'll make this a tool script so we can see the compositor effect work in editor.

Reviewer: ```suggestion
We'll make this a tool script so we can see the compositor effect work in the editor.
```

                    uniform.binding = 0
                    uniform.add_id(input_image)
                    var uniform_set = UniformSetCacheRD.get_cache(shader, 0, [ uniform ])

                    # Run our compute shader.
                    var compute_list := rd.compute_list_begin()
                    rd.compute_list_bind_compute_pipeline(compute_list, pipeline)
                    rd.compute_list_bind_uniform_set(compute_list, uniform_set, 0)
                    rd.compute_list_set_push_constant(compute_list, push_constant.to_byte_array(), push_constant.size() * 4)
                    rd.compute_list_dispatch(compute_list, x_groups, y_groups, z_groups)
                    rd.compute_list_end()

At the start of this method we check if we have a rendering device,
if our callback type is the correct one, and check if we have our shader.

.. note::

    The check for the effect type is only a safety mechanism.
    We've set this in our ``_init`` function, however it is possible
    for the user to change this in the UI.

Our ``p_render_data`` parameter gives us access to an object that holds
data specific to the frame we're currently rendering. We're currently only

> interested in our render scene buffers, which provides us access to all the

Reviewer: ```suggestion
interested in our render scene buffers, which provide us access to all the
```

internal buffers used by the rendering engine.
Note that we cast this to :ref:`RenderSceneBuffersRD <class_RenderSceneBuffersRD>`
to expose the full API to this data.

Next we obtain our ``internal size`` which is the resolution of our 3D render
buffers before they are upscaled (if applicable), upscaling happens after our
post processes have run.

From our internal size we calculate our group size, see our local size in our
template shader.

We also populate our push constant so our shader knows our size.
Godot does not support structs here **yet** so we use a
``PackedFloat32Array`` to store this data into. Note that we have

> to pad this array correctly.

Reviewer 2: I stumbled over this while writing my own effects.

You get a fairly cryptic error message when the push constants array isn't filled with a multiple of 4 floats / 16 bytes wide (?)... You counteracted this by adding a "reserved" vec2 above and filling it with zeroes in the gdscript part. Maybe elaborate on this specifically?

Author: @RedMser sadly it's far more complex then that, GPUs have some strict rules when it comes to memory layout of push constants and other such buffers, and it can differ depending on the type you specify. That's a subject that goes way beyond this documentation but I agree, should be something we should probably dictate its own chapter on.

Reviewer 2: Ah okay, in that case it's likely out of scope for this specific documentation...

Maybe the in-engine error message could be improved instead, IIRC it currently just says "pipeline requires a push constant to be set" which can be misleading. I can check details and open a bug ticket and/or a documentation / error message PR for this.

Author: I'm not sure, I'm also not 100% sure if we could properly check this as there will be no meta data on the struct to inform us of any misalignment. At the moment it purely checks if the expected size matches.

> We also populate our push constant so our shader knows our size.
> Godot does not support structs here **yet** so we use a
> ``PackedFloat32Array`` to store this data into. Note that we have
> to pad this array correctly.

Reviewer 3: ```suggestion
We also populate our push constant so our shader knows our size.
Godot does not support structs here **yet** so we use a
``PackedFloat32Array`` to store this data into. Note that we have
to pad this array with a 16 byte alignment. In other words, the
length of our array needs to be a multiple of 4.
```

Author: Owh thanks, that is definitely clearer. I know that for uniform buffers we have even stricter alignment rules but I wasn't sure how much applies to a push constant.

``NOTIFICATION_PREDELETE`` notification:

.. code-block:: gdscript

    # System notifications, we want to react on the notification that
    # alerts us we are about to be destroyed.
    func _notification(what):
        if what == NOTIFICATION_PREDELETE:
            if shader.is_valid():
                # Freeing our shader will also free any dependents such as the pipeline!
                RenderingServer.free_rid(shader)

Note that we do not use our mutex here even though we create our shader inside
of our render thread.
The methods on our rendering server are thread safe and ``free_rid`` will
be postponed cleaning up the shader until after any frames currently being
rendered are finished.

Also note that we are not freeing our pipeline. The rendering device does
dependency tracking and as the pipeline is dependent on the shader, it will
be automatically freed when the shader is destructed.

From this point onwards our code will run on the rendering thread.

Our next step is a helper function that will recompile the shader if the user
code was changed.

.. code-block:: gdscript

    # Check if our shader has changed and needs to be recompiled.
    func _check_shader() -> bool:

>         if !rd:

Reviewer: ```suggestion
        if not rd:
```
For all

        if !pipeline.is_valid():
            return false

        return true

At the top of this method we again use our mutex to protect accessing our
user shader code and our is dirty flag.
We make a local copy of the user shader code if our user shader code is dirty.

If we don't have a new code fragment, we return true if we already have a
valid pipeline.

If we do have a new code fragment we embed it in our template code and then
compile it.

.. warning::
    The code shown here compiles our new code in runtime.
    This is great for prototyping as we can immediately see the effect
    of the changed shader.

    This prevents precompiling and caching this shader which may be an issues
    on some platforms such as consoles.
    Note that the demo project comes with an alternative example where
    a ``glsl`` file contains the entire compute shader and this is used.
    Godot is able to precompile and cache the shader with this approach.

Finally we need to implement our effect callback, the rendering engine will call
this at the right stage of rendering.

.. code-block:: gdscript

    # Called by the rendering thread every frame.

>     func _render_callback(p_effect_callback_type, p_render_data):

Reviewer: ```suggestion
    func _render_callback(effect_callback_type, render_data):
```
General GDScript style

Author: Same reasoning as on the demo repo, this will cause a name clash with class properties.
